### PR TITLE
ValidatedQueries: start of validated queries API

### PR DESCRIPTION
**Pull request by jalevin**
**What this PR does / why we need it**:

Adds a basic API for validated queries (queries that are checked against what's saved in the dashboard). For now, it only makes sure that the provided dashboard and panel actually exist.

Relates to https://github.com/grafana/grafana-enterprise-partnerships-team/issues/67

### CODE REVIEW

File: pkg/api/metrics.go [hunk 1]
```diff
@@ -40,6 +42,80 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext) response.Response {
...
+		return false //response.Error(http.StatusBadRequest, "id is invalid", err)
+	}
+
+	query := models.GetDashboardQuery{
+		Id: id,
+	}
+
+	if err := bus.Dispatch(c.Req.Context(), &query); err != nil {
+		return false
+		//if errors.Is(err, models.ErrDataSourceNotFound) {
```
Comment (leandro-deveikis) [inline pkg/api/metrics.go:57]: Return same error as bus.Dispatch

File: pkg/api/metrics.go [hunk 2]
```diff
@@ -40,6 +42,80 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext) response.Response {
...
+		return response.Error(http.StatusBadRequest, "bad request data", err)
+	}
+
+	params := web.Params(c.Req)
+	dashboardId := params[":dashboardId"]
+	panelId := params[":panelId"]
+
+	if dashboardId == "" || panelId == "" {
+		// NOTE: Is this an appropriate status code?
+		return response.Error(http.StatusForbidden, "missing dashboard or panel ID", nil)
```
Comment (leandro-deveikis) [inline pkg/api/metrics.go:104]: Return BadRequest (400)

**Comment by jalevin** [comment]
Hey @kminehart would you be able to give us a review?

File: pkg/api/metrics.go [hunk 3]
```diff
@@ -40,6 +44,101 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext) response.Response {
...
+		//if errors.Is(err, models.ErrDataSourceIdentifierNotSet) {
+		//return response.Error(400, "Datasource id is missing", nil)
+		//}
+		//return response.Error(500, "Failed to query datasources", err)
+	}
+
+	// Dashboard has no properties. This would be weird, haven't written a test
+	// for this yet, but would most likely be a bug and return an error from
+	// bus.Dispatch
+	if query.Result == nil {
```
Comment (kylebrandt) [inline pkg/api/metrics.go:75]: Do we recover from panics in calling code here? Even if unlikely seems like the behavior of panicing the process would undesirable.

File: pkg/api/metrics.go [hunk 4]
```diff
@@ -40,6 +44,101 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext) response.Response {
...
+		return models.ErrDashboardIdentifierInvalid
+	}
+
+	query := models.GetDashboardQuery{
+		Id: id,
+	}
+
+	if err := bus.Dispatch(ctx, &query); err != nil {
+		return err
+		//if errors.Is(err, models.ErrDataSourceNotFound) {
```
Comment (kylebrandt) [inline pkg/api/metrics.go:63]: commened out code should be removed.

File: pkg/api/metrics.go [hunk 5]
```diff
@@ -40,6 +44,101 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext) response.Response {
...
+	id, err := strconv.ParseInt(dashboardId, 10, 64)
+	if err != nil {
+		return models.ErrDashboardIdentifierInvalid
+	}
+
+	query := models.GetDashboardQuery{
+		Id: id,
+	}
+
+	if err := bus.Dispatch(ctx, &query); err != nil {
```
Comment (kylebrandt) [inline pkg/api/metrics.go:61]: We are working on removing the bus. Can the SQL store be passed in instead (property on `hs *HTTPServer`) and its `GetDashboard` method be used instead?

File: pkg/api/metrics.go (hunk 3 above)
Comment (jalevin) [reply pkg/api/metrics.go:75]: We're not recovering from a panic here currently. Our thought process was that if this ever actually happens we have a serious problem and should shutdown. 

Happy to return an error and keep running if you think that makes more sense.

File: pkg/api/api.go [hunk 6]
```diff
@@ -381,6 +381,9 @@ func (hs *HTTPServer) registerRoutes() {
 		// DataSource w/ expressions
 		apiRoute.Post("/ds/query", authorize(reqSignedIn, ac.EvalPermission(ActionDatasourcesQuery)), routing.Wrap(hs.QueryMetricsV2))
 
+		// Validated query
+		apiRoute.Post("/dashboards/id/:dashboardId/panels/:panelId/query", authorize(reqSignedIn, ac.EvalPermission(ActionDatasourcesQuery)), routing.Wrap(hs.QueryMetricsFromDashboard))
```
Comment (kylebrandt) [inline pkg/api/api.go:385]: I think we generally prefer going with UIDs over IDs, it looks like GetDashboard can work by uid, do we need to be using id?

File: pkg/api/metrics.go (hunk 3 above)
Comment (kylebrandt) [reply pkg/api/metrics.go:75]: I think better to change it given that Grafana is still largely a monolith so recycle time can be a bit. And if there was a dashboard making a bunch of validate queries might take out multiple instances depending on the lb setup.

File: pkg/api/api.go (hunk 6 above)
Comment (jalevin) [reply pkg/api/api.go:385]: We didn't know that. Happy to make the change!

File: pkg/api/common_test.go [hunk 7]
```diff
@@ -337,21 +338,36 @@ func setupHTTPServer(t *testing.T, useFakeAccessControl bool, enableAccessContro
 	return setupHTTPServerWithCfg(t, useFakeAccessControl, enableAccessControl, cfg)
 }
 
+func setupHTTPServerWithMockDb(t *testing.T, useFakeAccessControl bool, enableAccessControl bool) accessControlScenarioContext {
```
Comment (ying-jeanne) [inline pkg/api/common_test.go:341]: nice

**Review by ying-jeanne — APPROVED** [review]
LGTM

**Pull request merged**